### PR TITLE
Fixed memory leak caused by empty await queues in WaitNotifyService. Awa...

### LINKED ISSUE #4432: Memory leak caused by empty await queues in WaitNotifyService
Memory leak when more then one thread try to lock on IMap key at the same time

tested on hazelcast version 3.4:

```java
        HazelcastInstance hz = Hazelcast.newHazelcastInstance();
        final IMap<Object, Object> testMap = hz.getMap("test");


        ExecutorService executorService = Executors.newFixedThreadPool(1);

        final AtomicInteger keySeq = new AtomicInteger(0);
        final AtomicBoolean done = new AtomicBoolean(false);

        while (true) {
            done.set(false);
            int uniquePart = keySeq.incrementAndGet();

            final int[] fatKey = new int[1000];
            fatKey[0] = uniquePart;

            testMap.lock(fatKey);


            executorService.execute(new Runnable() {
                @Override
                public void run() {

                    testMap.lock(fatKey);
                    testMap.unlock(fatKey);

                    done.set(true);
                }
            });

            // waiting second thread lock
            try {
                TimeUnit.MILLISECONDS.sleep(1);
            } catch (InterruptedException e) {
            }

            testMap.unlock(fatKey);

            while (!done.get()) {
                try {
                    TimeUnit.MILLISECONDS.sleep(1);
                } catch (InterruptedException e) {
                }
            }

        }
```

### PR DESCRIPTION
**Pull request by mdogan**
...it queues becoming empty should be removed from registration map. Fixes #4432.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/WaitNotifyService.java [hunk 1]
```diff
@@ -22,8 +22,35 @@
...
+     * returned from method {@link WaitSupport#getWaitKey()}.
+     * <p/>
+     * When operation is notified, it's re-executed by related scheduled mechanism.
+     * <p/>
+     * If wait time-outs, {@link WaitSupport#onWaitExpire()} method is called.
+     * <p/>
+     * This method should be called in the thread executes the actual {@link com.hazelcast.spi.WaitSupport} operation.
+     *
+     * @param waitSupport operation which will wait for notification
+     */
```
Comment (gurbuzali) [inline hazelcast/src/main/java/com/hazelcast/spi/WaitNotifyService.java]: :+1: good documentation

File: hazelcast/src/test/java/com/hazelcast/spi/impl/WaitNotifyServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        while (latch.getCount() > 0) {
+            LockSupport.parkNanos(1);
+            int awaitQueueCount = waitNotifyService.getAwaitQueueCount();
+            Assert.assertTrue(
+                    "Await queue count should be smaller than total number of threads: " + awaitQueueCount + " VS "
+                            + nThreads, awaitQueueCount < nThreads);
+        }
+    }
+
+    private static class InfinitelyLockAndUnlockTask implements Runnable {
```
Comment (sancar) [inline hazelcast/src/test/java/com/hazelcast/spi/impl/WaitNotifyServiceTest.java]: This class (Infinitely) is probably not doing what it says. :)

**Comment by sancar** [comment]
:+1:

File: hazelcast/src/test/java/com/hazelcast/spi/impl/WaitNotifyServiceTest.java (hunk 2 above)
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/spi/impl/WaitNotifyServiceTest.java]: Ah yes. It was before, then I changed the test but class name remained.

**Pull request merged**